File: python/src/lifegame/biosquare.py
from typing import Any, Generator, Protocol, Self

import numpy as np
from scipy.signal import convolve2d  # type: ignore

Generation = np.ndarray[Any, np.dtype[np.bool_]]
Biased = Generator[str, None, None]


class WorldCreator(Protocol):
    def create(self, nrows: int, ncols: int) -> Generation:
        pass


class LensFilter(Protocol):
    def observe(self, gen: Generation) -> Biased:
        pass


class BioSquare:
    KERNEL = np.array([[1, 1, 1], [1, -9, 1], [1, 1, 1]], dtype=np.int8)

    def __init__(
        self,
        nrows: int,
        ncols: int,
        world_creator: WorldCreator,
        lensfilter: LensFilter,
    ) -> None:
        self.generation = world_creator.create(nrows, ncols)
        self.creator = world_creator
        self.lensfilter = lensfilter

    def generate(self) -> Self:
        result: np.ndarray = convolve2d(
            self.generation,
            self.KERNEL,
            mode="same",
            boundary="wrap",
        )
        self.generation[result == 3] = True
        self.generation[(result != -6) & (result != -7) & (result < 0)] = False
        return self

    def observe(self) -> Biased:
        return self.lensfilter.observe(self.generation)

    def population_density(self) -> float:
        return self.generation.sum() / self.generation.size

    def reset(self) -> Self:
        self.generation = self.creator.create(*self.generation.shape)
        return self

File: rust/src/biosquare.rs
use crate::matrix::Matrix;
use rayon::prelude::{IntoParallelIterator, ParallelIterator};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};

pub type Generation = Matrix<bool>;
pub type Biased<'a> = Box<dyn Iterator<Item = String> + 'a>;

pub trait WorldCreator {
    fn create(&self, nrows: usize, ncols: usize) -> Generation;
}

pub trait LensFilter {
    fn observe<'a>(&'a self, gen: &'a Generation) -> Biased<'a>;
}

pub struct BioSquare {
    current: Generation,
    next: Mutex<Generation>,
    pub creator: Box<dyn WorldCreator>,
    pub lensfilter: Box<dyn LensFilter>,
}

impl BioSquare {
    pub fn new(
        nrows: usize,
        ncols: usize,
        world_creator: Box<dyn WorldCreator>,
        lensfilter: Box<dyn LensFilter>,
    ) -> Self {
        let current = world_creator.create(nrows, ncols);
        let next = Mutex::new(current.clone());
        Self {
            current,
            next,
            creator: world_creator,
            lensfilter,
        }
    }

    pub fn shape(&self) -> (usize, usize) {
        self.current.shape()
    }

    pub fn nrows(&self) -> usize {
        self.current.nrows()
    }

    pub fn ncols(&self) -> usize {
        self.current.ncols()
    }

    pub fn size(&self) -> usize {
        self.current.size()
    }

    pub fn generate(&mut self) -> &mut Self {
        (0..self.current.size()).into_par_iter().for_each(|i| {
            let (nrows, ncols) = self.current.shape();
            let row = i / ncols;
            let col = i % ncols;

            let mut count = 0;
            for row_offet in -1..=1 {
                for col_offset in -1..=1 {
                    if row_offet == 0 && col_offset == 0 {
                        continue;
                    }
                    if self.current[(
                        row.checked_add_signed(row_offet).unwrap_or(nrows - 1) % nrows,
                        col.checked_add_signed(col_offset).unwrap_or(ncols - 1) % ncols,
                    )] {
                        count += 1;
                    }
                }
            }

            let state = self.current[(row, col)];
            let mut next_generation = self.next.lock().unwrap();
            match (state, count) {
                (true, x) if x != 2 && x != 3 => next_generation[(row, col)] = false,
                (false, 3) => next_generation[(row, col)] = true,
                _ => (),
            }
        });
        self.next.lock().unwrap().overwrite(&mut self.current);
        self
    }

    pub fn observe(&self) -> Biased {
        self.lensfilter.observe(&self.current)
    }

    pub fn population_density(&self) -> f64 {
        let sum = Arc::new(AtomicUsize::new(0));
        let gen = &self.current;
        gen.par_for_each_index(|index| {
            if gen[index] {
                sum.fetch_add(1, Ordering::Relaxed);
            };
        });
        sum.load(Ordering::Relaxed) as f64 / gen.size() as f64
    }

    pub fn reset(&mut self) -> &mut Self {
        let (nrows, ncols) = self.current.shape();
        self.current = self.creator.create(nrows, ncols);
        self
    }
}
